plugin.video.catchuptvandmore/resources/lib/migration_utils.py

```diff
@@ -0,0 +1,174 @@
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License along
+    with Catch-up TV & More; if not, write to the Free Software Foundation,
+    Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+"""
+
+# The unicode_literals import only has
+# an effect on Python 2.
+# It makes string literals as unicode like in Python 3
+from __future__ import unicode_literals
+
+import os
+import json
+import importlib
+from builtins import str
+from builtins import range
+from kodi_six import xbmcvfs
+import xml.etree.ElementTree as ET
+
+
+from codequick import storage, Script, listing
+
+
+def migrate_from_pickled_fav():
+    """
+    This function moves existing pickled favourites in
+    the new json file used for the favourites.
+    The new format (json) appeared in 0.2.17~beta04
+    All user with version >= 0.2.17 will use favourites in the JSON format
+    Maybe we can remove the migration check on version 0.2.20?
+    """
+
+    # Move all pickled existing favs in json file
+    fav_pickle_fp = os.path.join(Script.get_info('profile'), "favourites.pickle")
+    if xbmcvfs.exists(fav_pickle_fp):
+        Script.log('Start favourites migration from pickle file to json file')
+        new_fav_dict = {}
+        with storage.PersistentDict("favourites.pickle") as db:
+            new_fav_dict = dict(db)
+        # Fix old fav
+        for item_hash, item_dict in new_fav_dict.items():
+            if 'params' in item_dict and isinstance(item_dict['params'], listing.Params):
+                new_fav_dict[item_hash]['params'] = dict(new_fav_dict[item_hash]['params'])
+                try:
+                    del new_fav_dict[item_hash]['params']['item_dict']['params']
+                except Exception:
+                    pass
+            if 'properties' in item_dict:
+                if isinstance(item_dict['properties'], listing.Property):
+                    new_fav_dict[item_hash]['properties'] = dict(new_fav_dict[item_hash]['properties'])
+        fav_json_fp = os.path.join(Script.get_info('profile'), "favourites.json")
+        with open(fav_json_fp, 'w') as f:
+            json.dump(new_fav_dict, f, indent=4)
+        xbmcvfs.delete(fav_pickle_fp)
+
+
+def migrate_old_menus_settings(menus_settings_fp):
+    """
+    This function moves old user settings concerning hidden menu
+    and items order in
+    the new json file used for that purpose.
+    The feature (json) appeared in 0.2.17~betaXX
+    All user with version >= 0.2.17 will use item settings in the JSON format
+    Maybe we can remove the migration check on version 0.2.20?
+    """
+
+    # If the json file already exists, the migration has already be done
+    if xbmcvfs.exists(menus_settings_fp):
+        return
+
+    # We directly read the XML setting file (faster than 'xbmcaddon.Addon().getSetting(s)')
```

Comment:
no you don't :) not allowed.

Reply:
Ha ha too bad 😛 
Do you know if there is a function in the API to get all the settings found in `settings.xml` and iterate on it?

Reply:
I don't know of any way of doing this for addons (only kodi settings). You can create an helper script that reads the settings.xml and dumps the content to some helper file (outside of kodi and don't submit it in this PR). Then in kodi you read that file and not the original one, using the proper api to change the settings.

Reply:
Done